File: src/PythonClient/lib/quixstreaming/builders/eventdefinitionbuilder.py
from quixstreaming import __importnet
import Quix.Sdk.Streaming

from ..models.eventlevel import EventLevel


class EventDefinitionBuilder(object):
    """
        Builder for creating event and group definitions for StreamPropertiesWriter
    """

    def __init__(self, net_object: Quix.Sdk.Streaming.Models.StreamWriter.EventDefinitionBuilder):
        """
            Initializes a new instance of EventDefinitionBuilder.

            Events:

            net_object (.net object): The .net object representing EventDefinitionBuilder.
        """

        self.__wrapped = net_object

    def set_level(self, level: EventLevel) -> 'EventDefinitionBuilder':
        """
        Set the custom properties of the event
        :param level: The severity level of the event
        :return: The builder
        """
        self.__wrapped = self.__wrapped.SetLevel(level.convert_to_net())
        return self

    def set_custom_properties(self, custom_properties: str) -> 'EventDefinitionBuilder':
        """
        Set the custom properties of the event
        :param custom_properties: The custom properties of the event
        :return: The builder
        """
        self.__wrapped = self.__wrapped.SetCustomProperties(custom_properties)
        return self

    def add_definition(self, event_id: str, name: str = None, description: str = None) -> 'EventDefinitionBuilder':
        """
        Add new event definition to the StreamPropertiesWriter. Configure it with the builder methods.
        :param event_id: The id of the event. Must match the event id used to send data.
        :param name: The human friendly display name of the event
        :param description: The description of the event
        :return: The builder
        """
        self.__wrapped = self.__wrapped.AddDefinition(event_id, name, description)
        return self

File: src/CsharpClient/Quix.Sdk.Streaming/Models/StreamWriter/EventDefinitionBuilder.cs
using Quix.Sdk.Process.Models;
using System;
using System.Collections.Generic;
using System.Text;

namespace Quix.Sdk.Streaming.Models.StreamWriter
{
    /// <summary>
    /// Builder for creating <see cref="EventDefinitions"/> within <see cref="StreamEventsWriter"/>
    /// </summary>
    public class EventDefinitionBuilder
    {
        private readonly StreamEventsWriter streamEventsWriter;
        private readonly string location;
        private Process.Models.EventDefinition properties;

        /// <summary>
        /// Initializes a new instance of <see cref="EventDefinitionBuilder"/>
        /// </summary>
        /// <param name="streamEventsWriter">Events writer owner</param>
        /// <param name="location">Location selected for the Event definition builder</param>
        /// <param name="properties">Events definition instance managed by the builder</param>
        public EventDefinitionBuilder(StreamEventsWriter streamEventsWriter, string location, Process.Models.EventDefinition properties = null)
        {
            this.streamEventsWriter = streamEventsWriter;
            this.location = location;
            this.properties = properties;
        }

        /// <summary>
        /// Set severity level of the Event
        /// </summary>
        /// <param name="level">The severity level of the event</param>
        public EventDefinitionBuilder SetLevel(EventLevel level)
        {
            if (this.properties == null)
            {
                throw new Exception($"You must add a definition before calling this method.");
            }

            this.properties.Level = level;

            return this;
        }

        /// <summary>
        /// Set custom properties of the Event
        /// </summary>
        /// <param name="customProperties"></param>
        public EventDefinitionBuilder SetCustomProperties(string customProperties)
        {
            if (this.properties == null)
            {
                throw new Exception($"You must add a definition before calling this method.");
            }

            this.properties.CustomProperties = customProperties;

            return this;
        }

        /// <summary>
        /// Add new Event definition, to define properties like Name or Level, among others.
        /// </summary>
        /// <param name="eventId">Event Id. This must match the event id you use to Event values</param>
        /// <param name="name">Human friendly display name of the event</param>
        /// <param name="description">Description of the event</param>
        /// <returns>Event definition builder to define the event properties</returns>
        public EventDefinitionBuilder AddDefinition(string eventId, string name = null, string description = null)
        {
            this.properties = this.streamEventsWriter.CreateDefinition(this.location, eventId, name, description);

            return this;
        }

    }

}
